pkg/cluster/cluster_test.go

```diff
@@ -0,0 +1,127 @@
+package cluster
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"context"
+	"errors"
+	"sort"
+
+	"reflect"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/improbable-eng/thanos/pkg/testutil"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+)
+
+func joinPeer(num int, knownPeers []string) (peerAddr string, peer *Peer, err error) {
+	port, err := testutil.FreePort()
+	if err != nil {
+		return "", nil, err
+	}
+	peerAddr = fmt.Sprintf("127.0.0.1:%d", port)
+	now := time.Now()
+	peerState1 := PeerState{
+		Type:    PeerTypeStoreSidecar,
+		APIAddr: fmt.Sprintf("sidecar-address:%d", num),
+		Metadata: PeerMetadata{
+			Labels: []storepb.Label{
+				{
+					Name:  "a",
+					Value: fmt.Sprintf("%d", num),
+				},
+			},
+			LowTimestamp:  timestamp.FromTime(now.Add(-10 * time.Minute)),
+			HighTimestamp: timestamp.FromTime(now.Add(-1 * time.Second)),
+		},
+	}
+
+	peer, err = Join(
+		log.NewNopLogger(),
+		prometheus.NewRegistry(),
+		peerAddr,
+		peerAddr,
+		knownPeers,
+		peerState1,
+		false,
+	)
+
+	return peerAddr, peer, nil
+}
+
+func sortStr(str []string) []string {
+	res := sort.StringSlice(str)
+	res.Sort()
+	return res
+}
+
+func TestPeers_PropagatingState(t *testing.T) {
+	ppOld := pushPullInterval
+	gOld := gossipInterval
+	// We need to decrease these options for quicker test. These are used in Join.
+	pushPullInterval = 100 * time.Millisecond
+	gossipInterval = 50 * time.Millisecond
+	defer func() {
+
+		pushPullInterval = ppOld
+		gossipInterval = gOld
+	}()
```

Comment:
As described above, I'd avoid this pattern of possible.

Reply:
make sense